File: abstract_classes/abstract_classes_java/implemented_subclasses_java/Queue.java
package abstract_classes.abstract_classes_java.implemented_subclasses_java;
import abstract_classes.abstract_classes_java.AbstractQueue;
import java.util.NoSuchElementException;

public class Queue extends AbstractQueue {
    private Node head;
    private Node tail;

    private class Node {
        int value;
        Node next;

        public Node(int value) {
            this.value = value;
            this.next = null;
        }
    }

    @Override
    public void enqueue(int value) {
        Node newNode = new Node(value);
        if (isEmpty()) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
        size++;
    }

    @Override
    public int dequeue() {
        if (isEmpty()) {
            throw new NoSuchElementException("Queue is empty");
        }
        int value = head.value;
        head = head.next;
        size--;
        if (isEmpty()) {
            tail = null;
        }
        return value;
    }
}

File: abstract_classes/abstract_classes_c++/implemented_subclasses_c++/Queue.cpp
#include <iostream>

namespace abstract_classes {
    class AbstractQueue {
    protected:
        int size;

    public:
        AbstractQueue() : size(0) {}

        virtual void enqueue(int value) = 0;
        virtual int dequeue() = 0;

        int getSize() {
            return size;
        }

        bool isEmpty() {
            return size == 0;
        }
    };
}

namespace concrete_classes {
    class Queue : public abstract_classes::AbstractQueue {
    private:
        int* elements;
        int front;
        int rear;
        int capacity;

    public:
        Queue(int capacity) : capacity(capacity), front(0), rear(-1) {
            elements = new int[capacity];
        }

        ~Queue() {
            delete[] elements;
        }

        void enqueue(int value) override {
            if (rear == capacity - 1) {
                std::cout << "Queue is full. Unable to enqueue." << std::endl;
                return;
            }

            elements[++rear] = value;
            size++;
        }

        int dequeue() override {
            if (isEmpty()) {
                std::cout << "Queue is empty. Unable to dequeue." << std::endl;
                return -1;
            }

            int value = elements[front++];
            size--;
            return value;
        }
    };
}